### PR TITLE
Add discovery_url

### PR DESCRIPTION
**Pull request by DrDaveD**
I made a mistake on specifying a specific token_url and it was hard to track down.  A discovery_url is harder to mess up if the token issuer supports the .well-known/openid-configuration.  The discovery url is also the type of url used by vault-plugin-auth-jwt so it works better for those of us who are using both plugins.

### CODE REVIEW

File: pkg/provider/basic.go [hunk 1]
```diff
@@ -159,12 +160,25 @@ func customFactory(vsn int, opts map[string]string) (Provider, error) {
...
 	}
 
-	authURL := opts["auth_code_url"]
+	var authURL string
+	var tokenURL string
+	discoveryURL := opts["discovery_url"]
+	if discoveryURL != "" {
+		provider, err := oidc.NewProvider(context.TODO(), discoveryURL)
+		if err != nil {
+			return nil, err
```
Comment (impl) [inline pkg/provider/basic.go:169]: You might want to wrap this in an `OptionError` like the others below.

**Review by impl — CHANGES_REQUESTED** [review]
Looks great; one small point of feedback on the error being returned. For the commit message, you'll want to use the ESLint format (e.g. `New: Add discovery_url`) also.

**Comment by DrDaveD** [comment]
I updated the commit message and added a Fix to use OptionError when NewProvider fails to work on the discovery_url option.

**Comment by impl** [comment]
One other thought I had after I stepped away from this for a second... would it make sense for the `auth_code_url` and `token_url` to override what comes back from discovery if you specify both, rather than being mutually exclusive? I'm not super familiar with OIDC Discovery so if that doesn't make any sense I'll be happy to get this in as-is. :) Thanks!

**Comment by DrDaveD** [comment]
No I don't think it makes a lot of sense to specify one or both of auth_code_url and token_url if discovery_url was specified.  I think people would only do that by mistake.  There is some overhead to use discovery_url, an extra round-trip, so I think if people want to specify the specific urls they shouldn't specify the discovery_url.

**Comment by impl** [comment]
Makes sense!

**Review by impl — APPROVED** [review]
LGTM

**Pull request merged**